docs/proposals/source-verificiation-policies.md

```diff
@@ -0,0 +1,313 @@
+
+### Overview
+
+Source verification policies are configured in the `sourceVerificationPolicies` field of an AppProject's spec. The field itself is a list, and you can configure an arbitrary amount of source verification policies.
+
+Each entry in the list has the following mandatory fields:
+
+* `repositoryPattern` defines a glob-style pattern that an Application's source repository URL must be matched against in order for the policy to apply
+* `verificationLevel` is the verification level for the policy
+* `verificationMethod` is the verification method for the policy
+* `trustedSigners` is the list of allowed signers
+
+### Anatomy of a source verification policy
+
+A source verification policy describes requirements for a source, as used by an Argo CD Application, that have to be met in order to be used for a sync.
+
+A source verification policy has the following properties:
+
+* A repository pattern (shell-glob type) that is matched against the URL of the source (mandatory)
+* A repository type that is matched against the application's repository type (mandatory)
+* A verification level that describes with which level the verification should be performed (mandatory)
+* A verification methods that describes how to the source should be verified (mandatory)
+* A list of identities of trusted signers (optional, if omitted, any known signer is trusted)
+
+Additionally, there are some optional properties you can set for a policy that depend on the verification level, method and respository type.
+
+Source verification policies are configured within an AppProject and should be considered an admin-level configurable.
+
+#### Repository pattern
+
+This is a glob-style pattern that is matched against the URL of the source to verify. When the pattern matches a source, it will be verified, otherwise verification of the source will be skipped.
+
+#### Repository type
+
+The type of repository. Right now, only "git" is supported.
+
+#### Verification levels
+
+Each source verification policy currently knows about four levels of verification:
+
+* `none`: As the name implies, verification will be disabled at this verification level.
+* `head`: This verification level will verify the commit that is pointed to by the HEAD of the target revision. In case target revision is an annotated tag, only the signature of the tag will be verified.
+* `progressive`: This verification level will verify all commits between the application's target revision and the revision it was last synced to, but not beyond that. The revision that the application is currently synced to does not necessarily need to be signed, so that a migration path from level `head` to the more secure method `progressive` is possible. However, if the application hasn't synced before (i.e. there is no previous revision in the application's sync history), the `progressive` level will inhibit the same behaviour as the `strict` verification level (see below).
```

Comment:
Should it be an explicit design goal to prevent downgrade attacks? e.g. a malicious user or git server could present HEAD as an older (signed) commit, pointing to an older, vulnerable version of the app? When syncing, argocd should only move to commits that are a descendent of the currently synced commit?

Reply:
Now I'm wondering if a malicious user in Git could modify the ancestry tree without affecting valid signatures on commits. 

The `strict` mode already checks the ancestry (i.e. only moves forward, not backwards). However, an attacker that has access to both, the Application CR and control over the Git server could potentially work around that.

Reply:
No, I think because the git signature covers both the contents of the commit and the digest of the parent commit, it's not possible to move commits around and retain a valid signature.

I think it's game over if an attacker can modify the Application CR or ArgoCD config in general. But I think that's fine - it would be nice to only have to trust ArgoCD and the cluster it runs on, but not the git server.

Reply:
> I think it's game over if an attacker can modify the Application CR

Can you please elaborate a little on this? The Application CR (as opposed to the AppProject CR) is eventually a user-modifiable resource, and a modification to it must not affect the security of the system.

Reply:
Sorry I wasn't very clear. If a user has a signing key, write access to any branch of a repo on the AppProject allowlist, and write access to the Application CR, they can get ArgoCD to deploy arbitrary resources (within the bounds of what the "best" AppProject allows as they can choose which one to be bound by). If people don't want that happen, I think the only solution is separation of duties so nobody has all 3 roles (or careful push rules on the git repos)